QEMU Morello

arichardson opened the pull request:
Supersedes #128 

Comments from https://github.com/CTSRD-CHERI/qemu/pull/148 also still need to be addressed, but otherwise it should mostly be ready for reviewing.

LawrenceEsswood committed:
Add sail parsing script and its output

LawrenceEsswood committed:
Add arm trickbox device

This is needed to run the Morello tests.

LawrenceEsswood committed:
Add get_capreg_or_special

LawrenceEsswood committed:
Changes to CHERI common code to support AARCH64

LawrenceEsswood committed:
Add a couple cap cursor modification helpers

LawrenceEsswood committed:
Fix otype sign related issues for AARCH64

LawrenceEsswood committed:
[Morello] DDC interposition depends on CCTLR_DDCBO

LawrenceEsswood committed:
Add TCG implementations of some capability operations

LawrenceEsswood committed:
Add basic CHERI support to AARCH64

LawrenceEsswood committed:
Add morello CPU type

LawrenceEsswood committed:
Add load/store pair and GCx family of instructions to morello

LawrenceEsswood committed:
Fix sail parser missing instructions

Squash me: fix decode script

LawrenceEsswood committed:
Add tcg condition to string function

LawrenceEsswood committed:
Add sync TCG instruction

LawrenceEsswood committed:
Add static tracking of capability decompression

LawrenceEsswood committed:
Give CapRegState a to string helper

LawrenceEsswood committed:
Make Morello specific changes to cheri utils

LawrenceEsswood committed:
Add another cap debug helper

LawrenceEsswood committed:
Lots more TCG helpers and Morello specific changes to setting offset and bounds

Co-authored-by: Alex Richardson <Alexander.Richardson@cl.cam.ac.uk>

LawrenceEsswood committed:
Pay attention to out of bounds exponants for bounds checks

LawrenceEsswood committed:
build/ccseal/copytype TCG for morello

LawrenceEsswood committed:
Better tracing of changes to registers

LawrenceEsswood committed:
Add restricted mode CSRS. PL_ALL redundant.

LawrenceEsswood committed:
General bug fixes for morello instructions

LawrenceEsswood committed:
Fix sign related issues with atomic operations

LawrenceEsswood committed:
Add helpers to get certain fields without decompression

LawrenceEsswood committed:
Allow atomics on Morello

LawrenceEsswood committed:
Add comments and some missing instructions to arm cheri decode

LawrenceEsswood committed:
Add some bit/byte packing functions

LawrenceEsswood committed:
64-bit memops should never have MO_SIGN

LawrenceEsswood committed:
Too lazy to fix, adding comment

LawrenceEsswood committed:
Add tag set many. Slightly refactor getting / setting multiple tags with atomicity in mind

LawrenceEsswood committed:
Remove some over-zelous asserts for morello

LawrenceEsswood committed:
Add some missing gpr_reg_modified calls

LawrenceEsswood committed:
Add CHERI MMU bits to morello

LawrenceEsswood committed:
Add exclusives and multi-tag loads/stores to morello

LawrenceEsswood committed:
TCG for bounds checking normal loads and stores

LawrenceEsswood committed:
Fix some capability checking logic for atomic operations

LawrenceEsswood committed:
Always bounds check with length at least 1

LawrenceEsswood committed:
Fix DDC exception number on MIPS

LawrenceEsswood committed:
Remove defunct print

LawrenceEsswood committed:
Assert too strict on AARCH64

LawrenceEsswood committed:
Fix usage of GETPC to only be at top level in helpers

LawrenceEsswood committed:
Define a separate aligned capability type

LawrenceEsswood committed:
Add load pair and branch and link instruction to morello.

LawrenceEsswood committed:
Add script to annotate instruction traces with source lines

LawrenceEsswood committed:
Improve implementation of sync

LawrenceEsswood committed:
Add userspace logging console command and remove translation cache flushing on mode change

LawrenceEsswood committed:
Add userspace logging to arm

LawrenceEsswood committed:
Enabled CHERI tags and logging on arm virt board

LawrenceEsswood committed:
Add debug assert for lazy capreg static optimisation

LawrenceEsswood committed:
Add some missing tcg tmp frees

LawrenceEsswood committed:
Fix some cursor sync issues

LawrenceEsswood committed:
Exceptions write to (and decompress) SP

LawrenceEsswood committed:
Cheri-fy arm's TPID registers

LawrenceEsswood committed:
Do CHERI things for DC ZVA

LawrenceEsswood committed:
Fix morello board minimum page size

LawrenceEsswood committed:
Restore state before CHERI exceptions on morello

arichardson committed:
Add a pytest script to run the morello tests

arichardson committed:
Fix out-of-bounds access found by UBSAN

disas_capreg_reset_all() will write past the end of the
cap_compression_states array.

arichardson committed:
Fix TCG temporary leak in gen_cap_get_type_for_copytype

arichardson committed:
Improve output for failed morello tests

arichardson committed:
disas/libvixl: Rename source files in preparation for update

This is needed in order to update to
https://github.com/Linaro/vixl/tree/morello which has disassembly support
for Morello.

arichardson committed:
Update disas/libvixl to https://github.com/Linaro/vixl/tree/morello

This includes disassembly support for Morello which makes -d instr
output a lot more useful.

arichardson committed:
[Morello] Forward PSTATE.C64 to the disassembler

This insures that mode-dependent instructions are decoded using the
capability registers instead of x registers.

arichardson committed:
[Morello] Implement -cheri-debugger-on-unrepresentable

arichardson committed:
[Morello] Fix out-of-bounds access when translating RRLEN

Passing AS_ZERO() to cpu_reg() will translate 31 to 32 which is not a
valid argument for cpu_reg (it uses 31 for the zero register).
Found while running the Morello tests with ASAN.

LawrenceEsswood committed:
Update generate_get_ddc_checked_gpr_plus_offset for Morello

LawrenceEsswood committed:
Add the morello config files

LawrenceEsswood committed:
Add morello board hardware

arichardson commented:
I think we should not use rebase+merge here but rather create a merge commit once all issues have been addressed. Rebasing and fixing history is rather awkward.

LawrenceEsswood committed:
[Morello] Always use soft-TLB for tag reads/writes

qwattash commented on accel/tcg/log_instr.c:
```diff
@@ -642,8 +645,6 @@ static void do_cpu_loglevel_switch(CPUState *cpu, run_on_cpu_data data)
     qemu_log_instr_loglevel_t next_level = data.host_int;
     bool next_level_active;
 
-    log_assert(qemu_loglevel_mask(CPU_LOG_INSTR));
```
Why was this removed? Is the assumption changed because of CPU_LOG_INSTR_U or was just redundant?

LawrenceEsswood replied:
The first I think? The assert was being hit on a normal path after the change.

qwattash commented on include/qemu/log.h:
```diff
@@ -64,6 +64,7 @@ static inline bool qemu_log_separate(void)
 /* LOG_STRACE is used for user-mode strace logging. */
 #define LOG_STRACE         (1 << 19)
 #define CPU_LOG_INSTR      (1 << 20)
+#define CPU_LOG_INSTR_U    (1 << 21)
```
Why this needed to be added among the top-level qemu tracing modes? My rationale for removing it a while back was that we would encode the mode per-cpu in the log state and when requested either by using NOPs or by using the qemu monitor we could call into the `cpu_loglevel_switch` functions.

LawrenceEsswood replied:
I was not aware do_cpu_loglevel_switch could be accessed from the qemu monitor. I added this pretty much because I wanted a way to enabled only userspace logging from the QEMU console / on startup with a command line argument.

qwattash replied:
Probably need to expose the function as it is currently internal. Leave it an I can change it later. I'm planning some work on the tracing side to support @heshamelmatary analysis anyway.

qwattash commented on accel/tcg/log_instr.c:
```diff
@@ -1036,6 +1065,319 @@ void qemu_log_instr_extra(CPUArchState *env, const char *msg, ...)
+
+TCGv_i64 qemu_log_printf_valid_entries;
+
+void qemu_log_printf_create_globals(void)
```
See comment on #148 May wan to move it to qemu_log_instr_init to limit the initialization hooks that we need to add to targets.

LawrenceEsswood replied:
Now somewhere more generic.

qwattash commented on accel/tcg/tcg-runtime.h:
```diff
@@ -148,12 +148,20 @@ GEN_ATOMIC_HELPERS(xchg)
 DEF_HELPER_3(pcc_check_load, cap_checked_ptr, env, tl, memop)
 /* Clear tags due to a store. Only calll this after the store succeeded. */
 DEF_HELPER_3(cheri_invalidate_tags, void, env, cap_checked_ptr, memop)
+/* Clear tags due to a store, the last argument is whether the store succeeded
+ */
```
```suggestion
/* Clear tags due to a store, the last argument is whether the store succeeded */
```

qwattash commented on target/arm/helper.c:
```diff
@@ -9851,15 +11110,36 @@ static void arm_cpu_do_interrupt_aarch64(CPUState *cs)
     }
 
-    pstate_write(env, PSTATE_DAIF | new_mode);
+#ifdef TARGET_CHERI
+    env->pstate &= ~PSTATE_C64;
+    if (cap_exception && (env->CCTLR_el[new_el] & CCTRL_C64E))
+        env->pstate |= PSTATE_C64;
+#endif
```
`pstate_write()` below will trash the C64 bit otherwise. The alternative is to mask the bit in `pstate_write` but I'm not sure whether it is relied upon in other places.
```suggestion
#ifdef TARGET_CHERI
    new_mode &= ~PSTATE_C64;
    if (cap_exception && (env->CCTLR_el[new_el] & CCTRL_C64E))
        new_mode |= PSTATE_C64;
#endif
```

LawrenceEsswood replied:
Oh oops... Something testing really should have found =/

arichardson replied:
Yeah we should definitely report this as missing coverage.

qwattash commented on accel/tcg/translator.c:
```diff
@@ -180,6 +205,13 @@ void translator_loop(const TranslatorOps *ops, DisasContextBase *db,
+#ifdef CONFIG_TCG_LOG_INSTR
+    /* Commit previous instruction */
+    if (unlikely(log_instr_enabled)) {
+        qemu_log_gen_printf_flush(db, true);
```
Is this duplicated here because when exiting the TB we will get the state cleared before we get to the commit in the next TB first instruction above? If so maybe warrant a comment.

LawrenceEsswood replied:
That is the case.

qwattash commented on hw/arm/Kconfig:
```diff
@@ -518,3 +518,6 @@ config ARMSSE_CPUID
     bool
+
+config MORELLO
+    bool
```
Missing EOF newline

qwattash commented on hw/arm/armtrickbox.c:
```diff
@@ -0,0 +1,149 @@
+static void empty_buffer(arm_trickbox_state *tb)
+{
+    tb->buf[tb->ndx] = '\0';
+    // Again, is there a QEMU way of doing this?
```
It might be `qemu_printf` which will emit to the current monitor or stdout?

qwattash commented on hw/arm/morello.c:
```diff
@@ -0,0 +1,154 @@
+
+    /* Start logging */
+#ifdef CONFIG_TCG_LOG_INSTR
+    qemu_log_instr_init(CPU(ARM_CPU(first_cpu)));
```
This should be called for each CPUState we created above.

LawrenceEsswood replied:
Done

qwattash commented on hw/arm/virt.c:
```diff
@@ -1673,6 +1683,11 @@ static void virt_cpu_post_init(VirtMachineState *vms, int max_cpus,
 
+    /* Start logging */
+#ifdef CONFIG_TCG_LOG_INSTR
+    qemu_log_instr_init(CPU(ARM_CPU(first_cpu)));
```
This should be called for each CPUState.

LawrenceEsswood replied:
Also done

qwattash commented on include/exec/log_instr_early.h:
```diff
@@ -0,0 +1,76 @@
+                        QEMU_LOG_PRINTF_BUF_DEPTH]; // arguments to printf calls
+    const char *fmts[QEMU_LOG_PRINTF_BUF_DEPTH];    // the printf fmts
+    uint64_t valid_entries; // bitmap of which entries are valid
+} qemu_log_printf_buf_t;
```
Since these types are not target-dependent you may get away with storing this in the `cpu_log_instr_state_t` in `include/qemu/log_instr.h` which is embedded in the `CPUState` struct directly, so you don't have to add it everywhere.

LawrenceEsswood replied:
I have done this, although it is a little unstable as it relies on the naming of members of different structs being the same. This _might_ be okay, as QEMU already have this requirement of naming a field 'parent_obj' anyway.

qwattash commented on scripts/sail_parser.py:
```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env python3
+
+# TODO: License
```
Add license before merging?

qwattash commented on target/arm/arm-semi.c:
```diff
@@ -303,13 +300,7 @@ static target_ulong arm_flen_buf(ARMCPU *cpu)
-    } else {
-        sp = env->regs[13];
-    }
+    target_ulong sp = arm_get_xreg(env, 31);
```
Isn't this getting the wrong register `31` vs `13` from the else branch?

LawrenceEsswood replied:
Yes, missed that it was different on the different branches.

qwattash commented on target/arm/cheri-archspecific-earlier.h:
```diff
@@ -0,0 +1,51 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
```
It is sad that we need 3 levels of early/earlier headers

LawrenceEsswood replied:
Wait till we need earliest!

qwattash commented on target/arm/cpu.h:
```diff
@@ -2402,10 +2506,17 @@ static inline uint64_t cpreg_to_kvm_id(uint32_t cpregid)
 #define ARM_CP_NO_GDB            0x4000
 #define ARM_CP_RAISES_EXC        0x8000
 #define ARM_CP_NEWEL             0x10000
+#define ARM_CP_CAP 0x20000
```
Whitespace as the others above?

qwattash commented on target/arm/helper.c:
```diff
@@ -4795,340 +4834,882 @@ static const ARMCPRegInfo v8_cp_reginfo[] = {
+     .access = PL2_RW | PL_IN_EXECUTIVE,
+     .type = ARM_CP_ALIAS,
+     .fieldoffset = offsetof(CPUARMState, sp_el[1]),
+     .type = ARM_CP_CAP_ON_MORELLO},
```
Double `.type` initialization?

qwattash commented on target/arm/helper.c:
```diff
@@ -5439,251 +6020,572 @@ static uint64_t cptr_el2_read(CPUARMState *env, const ARMCPRegInfo *ri)
+     .access = PL_IN_EXECUTIVE | PL3_RW,
+     .type = ARM_CP_ALIAS,
+     .fieldoffset = offsetof(CPUARMState, sp_el[2]),
+     .type = ARM_CP_CAP_ON_MORELLO},
```
Double `.type` initialization

qwattash commented on target/arm/translate-a64.c:
```diff
@@ -4167,6 +4619,7 @@ static void disas_add_sub_imm(DisasContext *s, uint32_t insn)
 
     TCGv_i64 tcg_rn = cpu_reg_sp(s, rn);
     TCGv_i64 tcg_rd = setflags ? cpu_reg(s, rd) : cpu_reg_sp(s, rd);
+
```
```suggestion
```

qwattash commented on target/arm/translate-a64.c:
```diff
@@ -4968,6 +5438,7 @@ static void disas_data_proc_3src(DisasContext *s, uint32_t insn)
         return;
     }
 
+
```
```suggestion
```

qwattash commented on target/arm/translate-a64.c:
```diff
@@ -5387,13 +5873,16 @@ static void handle_rev32(DisasContext *s, unsigned int sf,
                          unsigned int rn, unsigned int rd)
 {
     TCGv_i64 tcg_rd = cpu_reg(s, rd);
+
```
```suggestion
```

qwattash commented on target/arm/translate-a64.c:
```diff
@@ -5643,11 +6139,13 @@ static void handle_shift_reg(DisasContext *s,
 {
     TCGv_i64 tcg_shift = tcg_temp_new_i64();
     TCGv_i64 tcg_rd = cpu_reg(s, rd);
+
```
```suggestion
```

qwattash commented on target/arm/translate-cheri.inc.c:
```diff
@@ -0,0 +1,1802 @@
+    return false;
+}
+
+// Load/store common code
```
Easier to read if the docs for options are on different lines as a list.

qwattash commented on tcg/tcg.c:
```diff
@@ -4165,6 +4173,13 @@ int64_t tcg_cpu_exec_time(void)
+    TCGTemp *ts = arg_temp(op->args[0]);
+    tcg_debug_assert(ts->temp_global);
+    tcg_debug_assert(ts->val_type == TEMP_VAL_REG || ts->val_type == TEMP_VAL_MEM);
+    // Liveness analysis should sure that the sync happened at the last write
```
```suggestion
    // Liveness analysis should make sure that the sync happened at the last write
```

qwattash commented on target/arm/helper.c:
```diff
@@ -4795,340 +4834,882 @@ static const ARMCPRegInfo v8_cp_reginfo[] = {
-    { .name = "DC_CISW", .state = ARM_CP_STATE_AA64,
-      .opc0 = 1, .opc1 = 0, .crn = 7, .crm = 14, .opc2 = 2,
-      .access = PL1_W, .accessfn = access_tsw, .type = ARM_CP_NOP },
+    {.name = "IC_IALLUIS",
```
These need to check for the capability validity in C64. I think we are currently treating them as NOPs.

LawrenceEsswood replied:
I have just added a comment for now. I had already added a ARM_CP_IC_OR_DC type, but that checks only a single cache line. clear-all and the like probably need their access functions worked on, which will be slightly painful. As this is a NOP on QEMU, I will just leave it as a TODO.

qwattash commented on target/cheri-common/cheri_utils.h:
```diff
@@ -114,12 +130,22 @@ static inline bool cap_is_sealed_with_type(const cap_register_t* c) {
 static inline bool cap_is_in_bounds(const cap_register_t* c, target_ulong addr, size_t num_bytes) {
     cheri_debug_assert(num_bytes != 0);
+#ifdef TARGET_AARCH64
+    // Invalid exponant caps are always considered out of bounds.
```
```suggestion
    // Invalid exponent caps are always considered out of bounds.
```

qwattash commented on target/riscv/op_helper_cheri.c:
```diff
@@ -406,6 +406,12 @@ static target_ulong sc_c_impl(CPUArchState *env, uint32_t addr_reg,
     // in memory are the same as the ones that the load-reserved observed.
+    // FIXME: There is a bug here. If the MMU / Cap Permissions squash the tag,
+    // we may think the location has changed
+    // FIXME: when it has not. Use load_cap_from_memory_128_raw_tag to get the
```
I guess no need for the multiple `FIXME:`?

LawrenceEsswood committed:
Cache instructions should bounds check

LawrenceEsswood committed:
Fix C64 mode change for exception

LawrenceEsswood committed:
Fix morello 2 stage CHERI mmu permission checks

LawrenceEsswood committed:
Implement unprivileged morello loads/stores

LawrenceEsswood committed:
Configure morello cpu features

LawrenceEsswood committed:
[Morello] remove stray debug prints

LawrenceEsswood committed:
[Morello] Allow capability read/write functions CSRs

LawrenceEsswood committed:
sp is sometimes register 13

LawrenceEsswood committed:
Add missing newline

LawrenceEsswood committed:
Use qemu_printf

LawrenceEsswood committed:
Fix double field initialisation

LawrenceEsswood committed:
Fix comments and whitespace

LawrenceEsswood committed:
Add comment for cache maintenance instructions

LawrenceEsswood committed:
[LD/ST]CT should also do sp_alignment checks

LawrenceEsswood committed:
Fix off by one opc1 for CSCR_EL3

qwattash commented:
I think I added conflicts in TCG logging, should be just a couple of extra functions unrelated to this patch, so the resolution should be quick.

LawrenceEsswood committed:
Cannot use set tag at EL0

LawrenceEsswood committed:
Add CCTLR_EL12 to arm

LawrenceEsswood committed:
Dont unseal untagged sentries

LawrenceEsswood committed:
Generate trace for SCTAG

LawrenceEsswood committed:
Check SP alignment for LD|STCT in non C64 mode

LawrenceEsswood committed:
Take PC alignment exceptions on AARCH64

LawrenceEsswood committed:
Fix CHERI MMU bits in stage 2 translation

LawrenceEsswood committed:
Fix off by one in number of flags define

LawrenceEsswood committed:
Propagate a faulting address to pcc related faults

jrtc27 committed:
Remove old TARGET_CHERI128 config from morello-softmmu.mak

These got purged for CHERI-MIPS and CHERI-RISC-V during the removal of
lingering CHERI-256 support.

LawrenceEsswood committed:
Exception return should also clear tag if capability is sealed

jrtc27 committed:
Support GDB for Morello

LawrenceEsswood committed:
Propagate address for non-bounds faults for PCC and DDC

LawrenceEsswood committed:
TC TCPAC controls whether exceptions are enabled

LawrenceEsswood committed:
Print better register names when taking an exception

LawrenceEsswood committed:
Exception offset depends on executive bit

LawrenceEsswood committed:
Fix tracing EL

LawrenceEsswood committed:
Add better games for AARCH64 DBG registers

LawrenceEsswood committed:
Enable more breakpoints and watchpoints for morello

LawrenceEsswood committed:
Do not claim RAS is implemented on morello

LawrenceEsswood committed:
SPSel is RAZ/WI in restricted mode

LawrenceEsswood committed:
Capabilities enabled exceptions need to set pc properly

LawrenceEsswood committed:
Fix ignoring TGE when E2H enabled

LawrenceEsswood committed:
capability access SYN should have IL bit in a64

LawrenceEsswood committed:
Remove assert as behaviour is legal

LawrenceEsswood committed:
Make comment accurate

LawrenceEsswood committed:
Capability return without access system registers untags

LawrenceEsswood committed:
Correct ESR for SP alignment faults

LawrenceEsswood committed:
MSR immediate also needs to respect CHERI SPSEL rules

LawrenceEsswood committed:
Move PC alignment faults to tb_start to fix alignment priority

LawrenceEsswood committed:
Fix lots of unallocated morello encodings

LawrenceEsswood committed:
Take MMU protection faults for capability CAS even when not storing

LawrenceEsswood committed:
Add DBGCLAIM CSRS to make testhappy

LawrenceEsswood committed:
Probe with correct sizes so watchpoints trigger correctly

LawrenceEsswood committed:
Move some morello specific logic for capability jumps earlier

LawrenceEsswood committed:
Fix E2H related mistakes in FP enabled

LawrenceEsswood committed:
Fix tag setting enabled

LawrenceEsswood committed:
CAP_NO_SEALING is 64bit, not the size of the type field

LawrenceEsswood committed:
Do representability checks for legacy branches

LawrenceEsswood committed:
Only squash mutable permissions for tagged caps

LawrenceEsswood committed:
Generate correct ESR for MMU faults due to tags

LawrenceEsswood committed:
Add an always trapping version of PAGE_LC_TRAP

LawrenceEsswood committed:
Handle device memory correctly in conjunction with LC bit

LawrenceEsswood committed:
Fix cache maintenance instructions on morello

LawrenceEsswood committed:
CASL should not tag clear when compare fails

Co-authored-by: Alex Richardson <Alexander.Richardson@cl.cam.ac.uk>

LawrenceEsswood committed:
Representability check post-index for simd structure load/store

LawrenceEsswood committed:
Correctly set WnR bit for capability MMU faults

LawrenceEsswood committed:
Move TLB tag store faults into probe

LawrenceEsswood committed:
Sort fault priority of atomic cap operations

LawrenceEsswood committed:
Force morello to be little endian

LawrenceEsswood committed:
Handle jumping to non-sentry caps properly

LawrenceEsswood committed:
Morello set bounds immediate is exact

LawrenceEsswood committed:
Fix bounds setting on morello to do sign extention at correct point

LawrenceEsswood committed:
Fix access to sysregs requiring access sysreg permission

LawrenceEsswood committed:
Force printf commit for first instruction in a translation block

LawrenceEsswood committed:
Fix cheri bounds check size for multiple loads

LawrenceEsswood committed:
Expose symbol handling callback to arm boot

LawrenceEsswood committed:
Fix datasize in load/store exlusive pair

LawrenceEsswood committed:
CHERI bounds check needed on another cache op

LawrenceEsswood committed:
Fix cheri permission checks for double cap stores

arichardson committed:
Fail Morello tests with a more useful error message when misconfigured

Otherwise we just get a python length error instead of stdout+stderr.

arichardson committed:
Use the common tcg for setting merged registers to integers

arichardson committed:
Fix incorrect max_size argument in gen_mov_cap_select()

It needs to be a multiple of 32, not the actual copy size.

arichardson commented:
Rebased, but the individual commits will not build.

LawrenceEsswood committed:
Fix off by one on bounds check

arichardson commented:
rebased so that every commit builds for rv32+mips

LawrenceEsswood committed:
Move some aarch helpers from internal to cpu.h

LawrenceEsswood committed:
Implement alignment and morello exceptions

LawrenceEsswood committed:
Morello untags sealed capabilities just before branching to them

arichardson commented:
rebased, all commits after "Implement alignment and morello exceptions" build for morello as well.

jrtc27 commented on accel/tcg/log_instr.c:
```diff
@@ -739,18 +740,26 @@ static void do_global_loglevel_switch(CPUState *cpu, run_on_cpu_data data)
 
-    level = (request_stop) ? QEMU_LOG_INSTR_LOGLEVEL_NONE :
-        QEMU_LOG_INSTR_LOGLEVEL_ALL;
+    if(log_flags & CPU_LOG_INSTR_U) {
```
Needs some clang-format love for the series

arichardson replied:
Yeah and ideally checkpatch.pl too (see #169)

LawrenceEsswood committed:
Add TCG fast cap add for morello

LawrenceEsswood committed:
Respect exp and valid decode in more places in tcg

arichardson commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -973,6 +973,9 @@ static inline void gen_sp_set_decompressed_int(DisasContext *ctx, size_t offset)
                    offset + offsetof(cap_register_t, cr_bounds_valid));
 #endif
+    // Exponent
+    tcg_gen_movi_tl(temp, CC128_NULL_EXP);
```
This needs to be fixed for CHERI-64

arichardson committed:
Use CAP_CC() instead of hardcoding CC128

This should make it easier to adapt this code for 32-bit RISC-V.

LawrenceEsswood committed:
[Morello] Fix passing an invalid register number in BLR_BR_RET_CHKD

We have to use the raw register number not the one with the flags here.

Co-authored-by: Alex Richardson <Alexander.Richardson@cl.cam.ac.uk>

arichardson committed:
Change a few assertions to be debug-only

assert() is always on in QEMU, so change these to debug-only.

arichardson commented:
clang-formatted.

arichardson committed:
Make get_load_store_base_cap() always_inline

Running perf shows me that with the morello changes a RISC-V purecap
CheriBSD boot spends 4.7% of total time in this function.

Before this change:
```
hyperfine -L qemu 'qemu-system-riscv64cheri.morello,qemu-system-riscv64cheri.dev-pre-morello' '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd
_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/
alr48/cheri/output/sdk/bin/{qemu} --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest' -m 5
Benchmark #1: /home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest
  Time (mean ± σ):     221.103 s ±  2.055 s    [User: 208.764 s, System: 2.578 s]
  Range (min … max):   218.424 s … 223.736 s    5 runs

Benchmark #2: /home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.dev-pre-morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest
  Time (mean ± σ):     209.378 s ±  1.133 s    [User: 196.410 s, System: 3.062 s]
  Range (min … max):   207.857 s … 210.841 s    5 runs

Summary
  '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.dev-pre-morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest' ran
    1.06 ± 0.01 times faster than '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest'
```

After:
```
hyperfine -L qemu 'qemu-system-riscv64cheri.morello-ldst-reg-inlined,qemu-system-riscv64cheri.dev-pre-morello' '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/{qemu} --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest' -m 5
Benchmark #1: /home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.morello-ldst-reg-inlined --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest
  Time (mean ± σ):     206.196 s ±  1.410 s    [User: 194.016 s, System: 2.004 s]
  Range (min … max):   203.675 s … 206.932 s    5 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark #2: /home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.dev-pre-morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest
  Time (mean ± σ):     203.379 s ±  0.572 s    [User: 190.759 s, System: 2.685 s]
  Range (min … max):   202.715 s … 203.860 s    5 runs

Summary
  '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.dev-pre-morello --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest' ran
    1.01 ± 0.01 times faster than '/home/alr48/devel/cheribuild/test-scripts/run_cheribsd_tests.py --ssh-key /home/alr48/.ssh/insecure_id_ed25519.pub --architecture riscv64-purecap --kernel /local/scratch/alr48/cheri/output/rootfs-riscv64-purecap/boot/kernel.CHERI-PURECAP-QEMU/kernel --qemu-cmd /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.morello-ldst-reg-inlined --disk-image /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img --no-run-cheribsdtest'
```

arichardson commented on scripts/annotrace.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
```
This file should not be needed, we already have an annotate script. Unless this does something the existing one doesn't handle? @LawrenceEsswood

LawrenceEsswood replied:
Where is the alternative so I can compare?

arichardson replied:
scripts/symbolize-cheri-trace.py

LawrenceEsswood replied:
Looks to do roughly the same thing, minus a few small differences. symbolize-cheri-trace.py seems to use the wrong arguments for addr2line (which it uses if LLVM_SYMBOLIZER is not provided). Also, anno-trace goes and opens source files to get source annotations. Dunno if that is actually needed, or if I just didn't appreciate how much debug info was available through standard utilities.

arichardson commented:
rebased on top of #172

LawrenceEsswood committed:
Fix passing argument to handle_sys which does not uses 33 different register numbers

LawrenceEsswood committed:
Add PL_NO_SYSREG to special registers

arichardson commented:
I'm mostly fine with this PR as-is. There are a few things I would like to change, but rebasing this PR on top of those changes is painful, so I'd rather merge this first and then do the refactoring to remove all TARGET_AARCH64 from cheri-common.

@jrtc27 what do you think?

jrtc27 commented on tcg/tcg-op.c:
```diff
@@ -3432,64 +3546,65 @@ static void do_atomic_op_i64(TCGv_i64 ret, TCGv_cap_checked_ptr checked_addr,
-GEN_ATOMIC_HELPER(smax_fetch, smax, 1)
-GEN_ATOMIC_HELPER(umax_fetch, umax, 1)
+#define GEN_ATOMIC_HELPER(NAME, OP, NEW, SIGNED)                               \
+    static void *const table_##NAME[16] = {                                    \
```
Ideally this wouldn't be reformatted so it doesn't conflict horribly with upstream changes

arichardson committed:
Reduce diff to upstream by using g_assert_not_reached()

No need to depend on tcg_abort() here.

arichardson committed:
Avoid some large whitespace-only diffs to upstream

These were introduced by clang-format so this adds clang-format off
directives around the affected lines to avoid future churn.

arichardson committed:
Don't provide the csetflags helper for Morello

In Morello there are no flags in the metadata, the high address bits are
used for (software-defined) flags instead.

arichardson committed:
Reduce clang-format diffs to upstream

jrtc27 commented on accel/tcg/tcg-runtime.h:
```diff
@@ -148,12 +148,21 @@ GEN_ATOMIC_HELPERS(xchg)
 /* Same but relative to PCC */
 DEF_HELPER_3(pcc_check_bounds, void, env, tl, tl)
-/* Checks for loads relative to pcc */
-DEF_HELPER_3(pcc_check_load, cap_checked_ptr, env, tl, memop)
```
Needs rebasing? This shouldn't show up in the diff after #172, right?

jrtc27 commented on default-configs/targets/morello-softmmu.mak:
```diff
@@ -0,0 +1,7 @@
+
+INCLUDE_WORKAROUND=aarch64-softmmu.mak
+TARGET_CHERI=y
+TARGET_CHERI128=y
```
This no longer exists, we assume 2*TARGET_LONG_BITS is the capability size

jrtc27 replied:
(cherry-pick or fixup using https://github.com/CTSRD-CHERI/qemu/commit/f9957be3d1be751af8ac69e303683ce5cf594f97)

jrtc27 commented on default-configs/targets/morello-softmmu.mak:
```diff
@@ -0,0 +1,7 @@
+TARGET_CHERI=y
+TARGET_CHERI128=y
+TARGET_SUPPORTS_MTTCG=n
+TARGET_IS_MORELLO=1
```
Do we need the IS? We don't have one for TARGET_CHERI, nor does upstream have anything called TARGET_IS_FOO.

jrtc27 commented on disas/arm-a64.cc:
```diff
@@ -22,12 +22,14 @@ extern "C" {
 }
 
-#include "vixl/a64/disasm-a64.h"
+#include "vixl/aarch64/disasm-aarch64.h"
```
I'll note that, if we cared, we could upstream using a newer version of vixl, which would make our diff to use the Morello branch smaller. I wonder though if we should have a separate libvixl-morello that we point at so that we don't get any conflicts in the vixl source itself, and just need to change the build system plus this file (which would collapse to only adding the ISA option were upstream to update).

arichardson replied:
That would be nice, but I find submitting patches using git send-email rather awkward.

jrtc27 commented on hw/arm/armtrickbox.c:
```diff
@@ -0,0 +1,149 @@
+// The 'trickbox' device used by the test suite for Morello
+// Is an infinitely deep memory mapped fifo that will kill QEMU if \004 is sent
+// or otherwise print to the terminal There may be some other magic numbers this
+// accepts, I have not really reverse engineered the test cases yet.
```
```suggestion
// The 'trickbox' device used by the test suite for Morello
// is an infinitely deep memory mapped fifo that will kill QEMU if \004 is sent
// or otherwise printed to the terminal. There may be some other magic numbers this
// accepts, I have not really reverse engineered the test cases yet.
```
but could do with rewrapping

jrtc27 commented on hw/arm/armtrickbox.c:
```diff
@@ -0,0 +1,149 @@
+
+type_init(arm_trickbox_register)
+
+    static void arm_trickbox_mm_init(hwaddr base)
```
```suggestion
static void arm_trickbox_mm_init(hwaddr base)
```
clang-format doesn't like type_init I guess

jrtc27 commented on include/disas/dis-asm.h:
```diff
@@ -320,6 +320,7 @@ typedef struct disassemble_info {
   unsigned long flags;
 #define INSN_HAS_RELOC	0x80000000
 #define INSN_ARM_BE32	0x00010000
+#define INSN_ARM_C64	0x00000100
```
See the comment above it; technically this is in the wrong encoding space. Though actually this is just a relic from binutils, where this file originates, and doesn't really have much meaning here (INSN_HAS_RELOC isn't even set anywhere, only checked, so should be removed...).

Anyway, don't copy the binutils crap, go copy what's done for RISC-V (look for RISCV_DIS_FLAG).

arichardson replied:
You mean just use 0x1 and remove the comment above?

jrtc27 replied:
And move it out of the global header (and rename to ARM_DIS_FLAG_C64). These two only exist here with these names because they come from binutils.

jrtc27 commented on include/exec/cpu-all.h:
```diff
@@ -281,6 +281,13 @@ extern intptr_t qemu_host_page_mask;
+// Like PAGE_LC_TRAP but will also trap loads of untagged values
+#define PAGE_LC_TRAP_ANY 0x40000
+#define PAGE_C_BITS                                                            \
+    (PAGE_LC_CLEAR | PAGE_LC_TRAP | PAGE_LC_TRAP | PAGE_SC_CLEAR |             \
```
```suggestion
    (PAGE_LC_CLEAR | PAGE_LC_TRAP | PAGE_SC_TRAP | PAGE_SC_CLEAR |             \
```

LawrenceEsswood replied:
Good spot!

jrtc27 commented on include/exec/cpu-defs.h:
```diff
@@ -165,10 +167,16 @@ typedef struct CPUIOTLBEntry {
     /*
-     * ALL_ZERO_TAGBLK without either trapping or zero-ing is an invalid entry
-     * for tagmem_write and should cause a TLB miss if setting a tag to 1.
+     * There are two reasons to trap on writng a tag: first, if MMU protection
```
```suggestion
     * There are two reasons to trap on writing a tag: first, if MMU protection
```
though the grammar of this is bad, you can't split across two sentences like this; either merge them or swap the colon for a full stop and add a verb to the new sentence.

jrtc27 commented on include/exec/exec-all.h:
```diff
@@ -469,11 +469,13 @@ struct TranslationBlock {
-    (CF_COUNT_MASK | CF_LAST_IO | CF_USE_ICOUNT | CF_PARALLEL | CF_CLUSTER_MASK)
+#define CF_HASH_MASK                                                           \
+    (CF_COUNT_MASK | CF_LAST_IO | CF_USE_ICOUNT | CF_PARALLEL |                \
+     CF_CLUSTER_MASK | CF_LOG_INSTR)
```
```suggestion
     CF_LOG_INSTR | CF_CLUSTER_MASK)
```
to match the order they're defined in

jrtc27 commented on include/map.h:
```diff
@@ -0,0 +1,73 @@
+#define EVAL2(...) EVAL1(EVAL1(EVAL1(__VA_ARGS__)))
+#define EVAL3(...) EVAL2(EVAL2(EVAL2(__VA_ARGS__)))
+#define EVAL4(...) EVAL3(EVAL3(EVAL3(__VA_ARGS__)))
+#define EVAL(...)  EVAL4(EVAL4(EVAL4(__VA_ARGS__)))
```
EVAL does 243 expansions, this is horrendously inefficient. And of course MAP+EVAL are fundamentally broken due to having upper bounds. But this file doesn't look like it's used anywhere, so just remove this.

LawrenceEsswood replied:
It is needed for a secret commit (tm) that was removed until we get feedback from arm. But yes, it could be added when that commit does. I just want C to be functional, OK?

jrtc27 replied:
Hm, would rather we deal with it then and remove it for now

jrtc27 commented on include/tcg/tcg.h:
```diff
@@ -438,6 +438,33 @@ typedef enum {
+    case TCG_COND_NE:
+        return "!=";
+    case TCG_COND_LT:
+    case TCG_COND_LTU:
```
Do these not want disambiguating (e.g. as `<_u` and `<_s` like the operators in sail-riscv)

LawrenceEsswood replied:
Sounds sensible

jrtc27 commented on meson.build:
```diff
@@ -1,5 +1,5 @@
 project('qemu', ['c'], meson_version: '>=0.55.0',
-        default_options: ['warning_level=1', 'c_std=gnu99', 'cpp_std=gnu++11', 'b_colorout=auto'] +
+        default_options: ['warning_level=1', 'c_std=gnu99', 'cpp_std=gnu++14', 'b_colorout=auto'] +
```
Is this because of vixl?

arichardson replied:
Yeah, I got compiler errors otherwise...

jrtc27 commented on softmmu/vl.c:
```diff
@@ -1750,10 +1750,11 @@ static void version(void)
 #else
 #error "INVALID CHERI target"
 #endif
-    printf("Built with C0 as NULL register\n");
```
Was concerned this would break cheritest but seems it no longer looks for it

jrtc27 commented on target/arm/cheri-archspecific.h:
```diff
@@ -0,0 +1,234 @@
+
+#include "internals.h"
+
+typedef enum CheriCapExc {
```
Blegh this is a bit gross but fine, abstracting without using CHERI exception codes as the abstraction would be a lot of churn

jrtc27 commented on target/arm/cheri-archspecific.h:
```diff
@@ -0,0 +1,234 @@
+        target->cr_tag = 0;
+    }
+
+    // TODO: Reading BranchAddr in the arm ARM it looks like there is some
```
Is this still true? I'd expect there to be tests for that kind of thing

LawrenceEsswood replied:
There are are definitely some sign extensions that _should_ take place. I'm just unsure as to whether they are taking place elsewhere. If they are not, they should happen here.

jrtc27 commented on target/arm/cheri_helper.c:
```diff
@@ -0,0 +1,453 @@
+void helper_swap_cap_via_cap(CPUArchState *env, uint32_t cd, uint32_t cs,
+                             uint32_t cb, target_ulong addr)
+{
+    // Yes, I really meant to swap cs/cd here.
```
Maybe call them cs and ct like in the manual? cd implies destination, which is not what it is, SWP's arguments are backwards from what you would think they should be (and compared to CAS). And that fact should be added as an explanation in the comment.

jrtc27 commented on target/arm/cheri_helper.c:
```diff
@@ -0,0 +1,453 @@
+    bool tag = load_cap_from_memory_raw(env, &pesbt, &cursor, cb, cbp, addr,
+                                        _host_return_address, NULL);
+
+    bool do_store = !compare;
+
+    if (compare) {
+        do_store = (cursor == get_without_decompress_cursor(env, cs)) &&
+                   (pesbt == get_without_decompress_pesbt(env, cs)) &&
+                   (tag == get_without_decompress_tag(env, cs));
+    }
```
This isn't a great way to write this, make it an if/else, harder to follow as it stands

LawrenceEsswood replied:
Sure

jrtc27 commented on target/arm/cheri_helper.c:
```diff
@@ -0,0 +1,453 @@
+        CAP_cc(update_otype)(&base, CAP_OTYPE_UNSEALED);
+    }
+
+#if (KEEP_BRANCH_AND_LINK_BUG)
```
Is this for the old ASL? Do we need this?

LawrenceEsswood replied:
Looks BLR (memory indirect) has been fixed, so can kill.

jrtc27 commented on target/arm/cheri_helper.c:
```diff
@@ -0,0 +1,453 @@
+
+    cap_register_t base = *get_capreg_or_special(env, cn);
+
+    tags &= (1 << 4) - 1;
```
Where's the `CapIsSystemAccessEnabled() && !IsTagSettingDisabled()` requirement that forces it to 0 otherwise?

LawrenceEsswood replied:
In translate-cheri.inc.c. Done at translate time because all those bits are cached.

arichardson replied:
Probably makes sense to add a debug assertion?

LawrenceEsswood replied:
Sure

jrtc27 commented on target/arm/cheri_helper.c:
```diff
@@ -0,0 +1,453 @@
+
+    uint32_t perms = CAP_PERM_STORE;
+    if (tags)
+        perms |= CAP_PERM_STORE_CAP;
```
This isn't quite right, the ASL does the tags one by one, so STCT of 2 via a non-STORE_CAP capability will clear the first capability tag and fault trying to write the second. I don't know if that is _required_ behaviour or if implementations are permitted to make it a single atomic store (I would hope so otherwise it's not very accelerating?) and just a quirk of the ASL.

Of course such problems go away if you remove STCT from the architecture and just have a CClearTags...

LawrenceEsswood replied:
You are right. The question is, do we slow down emulation to correct this behaviour nobody is ever going to hit outside of a test for exactly this?

jrtc27 replied:
Let's go ask Arm whether the hardware takes that liberty... :)

LawrenceEsswood replied:
Looks like its okay

jrtc27 commented on target/arm/cpu.c:
```diff
@@ -65,7 +90,8 @@ static void arm_cpu_synchronize_from_tb(CPUState *cs,
      */
     if (is_a64(env)) {
-        env->pc = tb->pc;
+        // LETODO: I dont know if this needs bounds checking
```
Is this still unknown?

LawrenceEsswood replied:
Yup =/. All this magic PC re-sync scares me. My hunch is no because we regen code if its called with different bounds?

jrtc27 commented on target/arm/cpu.c:
```diff
@@ -128,7 +154,12 @@ static void cp_reg_reset(gpointer key, gpointer value, gpointer opaque)
+        CPREG_FIELDCAP(&cpu->env, ri) = ri->capresetvalue;
+    } else
+#endif
+        if (cpreg_field_is_64bit(ri)) {
```
Was this clang-format or -Wmisleading-indentation?

arichardson replied:
clang-format.

jrtc27 commented on target/arm/cpu.c:
```diff
@@ -217,7 +248,14 @@ static void arm_cpu_reset(DeviceState *dev)
+
+#ifdef TARGET_CHERI
+        reset_capregs(env);
+        env->pc.cap = cc128_make_max_perms_cap(0, cpu->rvbar, CAP_MAX_TOP);
```
```suggestion
        env->pc.cap = CAP_cc(make_max_perms_cap)(0, cpu->rvbar, CAP_MAX_TOP);
```
so it doesn't need to be changed once we make cc128 and "cc128m" separate

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -218,6 +224,18 @@ typedef struct ARMPACKey {
 
+#ifdef TARGET_CHERI
+#include "cheri-lazy-capregs-types.h"
+#define AARCH_REG_TYPE aligned_cap_register_t
```
Is there a reason not to typedef it? I don't see anything relying on the deferred evaluation of macro bodies.

LawrenceEsswood replied:
No reason I can see

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -1083,11 +1162,18 @@ int fp_exception_el(CPUARMState *env, int cur_el);
 
+#ifdef TARGET_CHERI
+// Morello does not support 32-bit, so might as well optimise everything away
+#define is_a64(env) 1
```
Might be better to just push it into the function as `return 1` and mark the argument as possibly-unused, otherwise callers _might_ start to get spurious unused argument warnings. Should optimise to the same thing.

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -1201,11 +1287,24 @@ void pmu_init(ARMCPU *cpu);
 
 #define CPTR_TCPAC    (1U << 31)
-#define CPTR_TTA      (1U << 20)
+#define CPTR_TTA (1U << 20) /* CPTR_EL3 */
```
clang-format strikes again

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -1201,11 +1287,24 @@ void pmu_init(ARMCPU *cpu);
 #define CPTR_TZ       (1U << 8)   /* CPTR_EL2 */
 #define CPTR_EZ       (1U << 8)   /* CPTR_EL3 */
 
+#define CPTR_TTA_EL2 (1U << 28)
```
Line these up like the others

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -1301,6 +1400,9 @@ void pmu_init(ARMCPU *cpu);
 #define PSTATE_UAO (1U << 23)
 #define PSTATE_TCO (1U << 25)
+// Morello documentation seems to indicate this was the correct bit in SPSR for
+// PSTATE.C64
```
Not sure why this is here, it's clearly defined to be that in the revised definitions of SPSR_ELx

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -2063,6 +2168,10 @@ uint64_t arm_hcr_el2_eff(CPUARMState *env);
+#ifdef TARGET_CHERI
+    // Morello always a64
+    return true;
+#endif
```
I'd `#else ... #endif` the function, might get dead code warnings otherwise...

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -2768,6 +2937,24 @@ static inline bool cpreg_field_is_64bit(const ARMCPRegInfo *ri)
+static inline bool cpreg_field_is_cap(const ARMCPRegInfo *ri)
+{
+#ifdef TARGET_CHERI
+    return !!(ri->type & ARM_CP_CAP);
```
Inconsistent about !! vs (A & B) == A (as done below)

LawrenceEsswood replied:
They are semantically different, as ARM_CP_CAP_ONLY is not a single bit. I will make this the same as below anyway as for the single bit it is the same, and it may change in the future.

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -218,6 +224,18 @@ typedef struct ARMPACKey {
+#endif
 
 typedef struct CPUARMState {
     /* Regs for current mode.  */
```
Should `regs` be ifdef'd out?

LawrenceEsswood replied:
Well, morello does not support 32 bit mode, but sadly this can't be ifdef'd out without a factor / an ugly define hack. There is lots of 32-bit supporting code that will not compile if that array is not there, even though it well never get used in practice.

jrtc27 replied:
:( It's a shame upstream can't just operate on the 64-bit registers in 32-bit mode (with maybe a special register for PC), then it wouldn't be such an issue

LawrenceEsswood replied:
A comment there says "32/64 switch only happens when taking and returning from exceptions so the overlap semantics are taken care of then instead of having a complicated union". Apparently whoever wrote that thought that remembering to sync state on every exception / return is somehow less complicated than typing the word union?

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -3194,7 +3381,11 @@ static inline bool arm_cpu_data_is_big_endian_a32(CPUARMState *env,
 
 static inline bool arm_cpu_data_is_big_endian_a64(int el, uint64_t sctlr)
 {
+#ifdef TARGET_CHERI
```
Is this really needed? The fields should be constant, and I doubt it's a hot path.

LawrenceEsswood replied:
I think constant is valid, but QEMU does not enforce it currently. arm ARM says if an endianess is not supported the E field is RES0/RES1 as is appropriate.

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -3242,6 +3504,43 @@ FIELD(TBFLAG_ANY, FPEXC_EL, 22, 2)
+FIELD(TBFLAG_CHERI, SYSTEM, 8, 1)     // pcc.perms.system
+FIELD(TBFLAG_CHERI, SETTAG, 9, 1)
+FIELD(TBFLAG_CHERI, CAP_ENABLED, 10,
+      1) // 1 if capability instructions are trapped
```
clang-format?... :(

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -3979,20 +4344,149 @@ static inline bool isar_feature_any_tts2uxn(const ARMISARegisters *id)
+    if (cpu_mmu_index(env, 0) == ARMMMUIdx_Stage2) {
+        ttbr = env->cp15.vttbr_el2;
+    } else if (1) {
+        return env->cp15.ttbr0_el[arm_current_el(env)];
```
`ttbr = `, surely?

LawrenceEsswood replied:
Seems likely.

jrtc27 commented on target/arm/cpu.h:
```diff
@@ -3979,20 +4344,149 @@ static inline bool isar_feature_any_tts2uxn(const ARMISARegisters *id)
+    // TODO: This is slightly broken because tbbrn is defaulting to 0
+    if (cpu_mmu_index(env, 0) == ARMMMUIdx_Stage2) {
+        ttbr = env->cp15.vttbr_el2;
+    } else if (1) {
```
I assume this is the brokenness...

LawrenceEsswood replied:
Right, this is the brokenness. Not sure what to do here. This function is just used for logging. Is the ASID meant to be for the current PC? Seems like the most sensible thing. We could pass it through for all targets, even though arm is the only target where ASID depends an address.

jrtc27 replied:
Yes, it's just for logging a bit of additional info next to the current instruction so it's easier to trace multiple processes. That might be the most sensible thing to do if you need to pick a single ASID; it'll be a little confusing as you cross between kernel and userspace, but I think it's about the best you can do.

jrtc27 commented on target/arm/cpu64.c:
```diff
@@ -728,10 +867,11 @@ static void aarch64_max_initfn(Object *obj)
+    {.name = "cortex-a53", .initfn = aarch64_a53_initfn},
+    {.name = "cortex-a72", .initfn = aarch64_a72_initfn},
+    {.name = "morello", .initfn = aarch64_morello_initfn},
+    {.name = "max", .initfn = aarch64_max_initfn},
```
clang-format off

jrtc27 commented on target/arm/gdbstub64.c:
```diff
@@ -20,20 +20,22 @@
 #include "cpu.h"
 #include "exec/gdbstub.h"
 
+// TODO: CHERI-fy
```
cherry-pick https://github.com/CTSRD-CHERI/qemu/commit/3cc9ca36ae39fb6a7a3f640031acd905f26b7306 :)

jrtc27 commented on target/arm/helper-a64.c:
```diff
@@ -1077,15 +1118,20 @@ void HELPER(exception_return)(CPUARMState *env, uint64_t new_pc)
     env->pstate |= PSTATE_IL;
-    env->pc = new_pc;
+    // LETODO
+    ASSERT_IF_CHERI();
```
Is this one of the things that's been fixed after running the ACK but not pushed to this particular branch?

LawrenceEsswood replied:
Nope. Nothing has used this yet! I suppose if I fixed more ACK things it might eventually come up.

jrtc27 commented on target/arm/helper.c:
```diff
@@ -4273,6 +4393,12 @@ static uint64_t mpidr_read_val(CPUARMState *env)
         if (cpu->mp_is_up) {
             mpidr |= (1u << 30);
         }
+#define SUPPORT_ACK
```
Doesn't belong here. Maybe belongs as a run-time property set by hw/arm/morello.c?

LawrenceEsswood replied:
You are right, it does not. It needs plumbing, probably with the secret commit (tm).

jrtc27 commented on target/arm/helper.c:
```diff
@@ -4794,27 +4930,28 @@ static void sdcr_write(CPUARMState *env, const ARMCPRegInfo *ri,
 static const ARMCPRegInfo v8_cp_reginfo[] = {
     /* Minimal set of EL0-visible registers. This will need to be expanded
      * significantly for system emulation of AArch64 CPUs.
+     * clang-format off
```
Put this as a separate comment above the definition, or make this a separate one-line comment here and move the off within the outer array so they match

jrtc27 commented on target/arm/helper.c:
```diff
@@ -5439,251 +5579,571 @@ static uint64_t cptr_el2_read(CPUARMState *env, const ARMCPRegInfo *ri)
-      .opc0 = 3, .opc1 = 4, .crn = 10, .crm = 3, .opc2 = 0,
-      .access = PL2_RW, .type = ARM_CP_CONST,
-      .resetvalue = 0 },
+    {.name = "HCR_EL2",
```
clang-format off for a bunch of these

jrtc27 commented on target/arm/helper.c:
```diff
@@ -8083,14 +8752,22 @@ void register_cp_regs_for_features(ARMCPU *cpu)
-              .resetvalue = 0 },
-            REGINFO_SENTINEL
-        };
+            {.name = "VBAR",
```
clang-format off, just rework writefn to be on its own line

jrtc27 commented on target/arm/helper.c:
```diff
@@ -8189,6 +8866,184 @@ void register_cp_regs_for_features(ARMCPU *cpu)
+    ARMCPRegInfo cheri_regs[] = {
+        // We swap the relevent DDC into this register, so it is always
+        // accessible
+        {.name = "DDC",
```
Would be nicer to format these in the more compact style used elsewhere

arichardson replied:
Not sure if there is a clang-format option to do that, but I think we can deal with it when we merge from upstream. Reverting the formatting while keeping the changes is really tedious.

jrtc27 replied:
This one I care about less because it's never going to conflict, though should still be done in the same style as the others. I do care about the other big tables, it makes it impossible to read the diff.

jrtc27 commented on target/arm/mte_helper.c:
```diff
@@ -821,7 +821,8 @@ uint64_t HELPER(mte_checkN)(CPUARMState *env, uint32_t desc, uint64_t ptr)
  * Perform an MTE checked access for DC_ZVA.
  */
-uint64_t HELPER(mte_check_zva)(CPUARMState *env, uint32_t desc, uint64_t ptr)
+target_ulong HELPER(mte_check_zva)(CPUARMState *env, uint32_t desc,
```
Should really be uint64_t, the argument is and it gets returned as-is. This is an A64-only helper so target_ulong will be uint64_t.

jrtc27 commented on target/arm/translate-a64.c:
```diff
@@ -1871,50 +2361,54 @@ static void handle_sys(DisasContext *s, uint32_t insn, bool isread,
     case ARM_CP_DC_GVA:
         {
-            TCGv_i64 clean_addr, tag;
+        TCGv_i64 tag;
```
clang-format is wrong here

arichardson replied:
Weird upstream formatting IMO. If the braces are part of the case you shouldn't indent it again...

arichardson replied:
Aha `IndentCaseBlocks: true` seems to fix it.

jrtc27 commented on target/arm/translate-a64.c:
```diff
@@ -1816,6 +2243,45 @@ static void handle_sys(DisasContext *s, uint32_t insn, bool isread,
 
+#ifdef TARGET_CHERI
+    // Morello versions target the entire capability sized register
+    bool is_morello = extract32(insn, 25, 4) == 0x1;
```
Really is_cap not is_morello...

jrtc27 commented on target/arm/translate-a64.c:
```diff
@@ -1933,7 +2427,49 @@ static void handle_sys(DisasContext *s, uint32_t insn, bool isread,
+        // it should be ensured they do not cause out of bounds.
+        if (isread) {
+            if (ri->type & ARM_CP_CONST) {
+                assert(0 && "TODO");
```
Is this "TODO" or "No constant capability system registers exist"?

LawrenceEsswood replied:
Both?

jrtc27 replied:
As in, is it a property of the architecture that no such registers exist? If so the assertion should state that, otherwise it looks like we're missing something (and doesn't hint that if this assertion is hit something else has gone wrong rather than this line).

jrtc27 commented on target/arm/translate-a64.c:
```diff
@@ -14633,8 +15361,17 @@ static void disas_a64_insn(CPUARMState *env, DisasContext *s)
+    // Top 4 bits of op0
     switch (extract32(insn, 25, 4)) {
-    case 0x0: case 0x1: case 0x3: /* UNALLOCATED */
+    case 0x1: /* Morello Instructions */
```
Swap these two cases around (to match 0x2)

LawrenceEsswood replied:
Not sure which cases you mean. Currently, 0x1 will fall through to unallocated_encoding, which is nice if target is not CHERI.

jrtc27 replied:
Oh I see, I missed that in my pass through, could do with a `/* fallthrough */` comment. Though I'd be inclined to just rewrite it as:

```
case 0x1:
#ifdef TARGET_CHERI
    if (!disas_cheri(s, insn))
#endif
    {
        unallocated_encoding(s);
    }
    break;
```
and not need the fallthrough, then you can move this case to be next to 2 rather than first here.

jrtc27 commented on target/arm/translate-cheri.inc.c:
```diff
@@ -0,0 +1,2045 @@
+    }
+
+    if (c64) {
+        // Derive a capability
```
Does there need to be a capabilities enabled check for C64? There are various ways you could end up in C64 with capabilities disabled and the next instruction being ADR(D)P, unless C64 makes you trap if capabilities are disabled (which would be somewhat sensible, though technically you could stick to all the integer instructions...)?

LawrenceEsswood replied:
There is no CheckCapabilitiesEnabled() in the ASL. I assume this is the case because ADR(D)P are not actually in the morello encoding space, even if they claim to be?

jrtc27 replied:
Hm, that seems like a spec bug, unless it's implicitly done elsewhere. As things stand I believe this is a half-hole in the spec; in practice it doesn't matter because you'd have to ERET with SPSR.C64 set (`BX #4` _is_ disabled if capabilities aren't enabled, so userspace can't get into C64 by itself), but it technically allows you to get into a state where it's possible to write back a capability when you shouldn't be able to.

jrtc27 replied:
Ok, R_VCNGF is the important rule, it's impossible to end up in C64 mode with capability accesses trapped (and ERET will force PSTATE.C64 to 0 regardless of SPSR_ELx.C64 in that case). Guess I should check you've done that... :)

LawrenceEsswood replied:
is_access_to_capabilities_enabled_at_el is used for ERET, which is copied verbatim from the ASL and will make sure SPSR_C64 is 0. So there is certainly no way of having C64 on ERET if caps not enabled. The other instructions that could set it all do a CheckCapabilitiesEnabled. Think we are OK.

jrtc27 replied:
:+1: so close to the prize of finding a subtle hole in the architecture..

jrtc27 commented on target/arm/translate-cheri.inc.c:
```diff
@@ -0,0 +1,2045 @@
+        tcg_temp_free_i32(target_regnum);
+        tcg_temp_free_i32(link_regnum);
+
+        // LETODO: Check this.
```
Has this been checked?

jrtc27 commented on target/arm/translate-cheri.inc.c:
```diff
@@ -0,0 +1,2045 @@
+{
+
+    // This instruction has really weird treatment of C31 in the ASL.
+    // For the purposes of loading, C31 is CSP. For storing it is 0.
```
The ASL never writes to C[n] as such? It always* writes back C[30] as the link register, and writes the (possibly unsealed) base to C[29] if n == 29.

Maybe an older version of the ASL erroneously always wrote back but that got changed?

\* Except for n == 30 when it's constrained unpredictable, but still either writes UNKNOWN to C[30] or is a complete NOP

jrtc27 commented on target/arm/translate-cheri.inc.c:
```diff
@@ -0,0 +1,2045 @@
+    if (a->opc == 0) {
+        helper = &gen_helper_csetbounds;
+    } else if (a->opc == 1) {
+        helper = &gen_helper_csetboundsexact;
```
```suggestion
    if (!(a->opc & 2)) {
        cheri_cap_cap_int_helper *helper;
        if (a->opc == 0) {
            helper = &gen_helper_csetbounds;
        } else {
            helper = &gen_helper_csetboundsexact;
        }
        return gen_cheri_cap_cap_int(ctx, a->Cd, a->Cn, a->Rm, helper);
    }
```
would make the control flow easier to follow (the `if (helper)` at the end is a bit weird)

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -395,3 +588,1661 @@ static inline void gen_check_cond_branch_target(DisasContext *ctx,
+{
+#if MERGED_FILE
+    if (!lazy_capreg_number_is_special(regnum))
+        tcg_gen_sync_i64(target_get_gpr_global(ctx, regnum));
```
```suggestion
        tcg_gen_sync_tl(target_get_gpr_global(ctx, regnum));
```

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -395,3 +588,1661 @@ static inline void gen_check_cond_branch_target(DisasContext *ctx,
+{
+#if MERGED_FILE
+    if (!lazy_capreg_number_is_special(regnum))
+        tcg_gen_discard_i64(target_get_gpr_global(ctx, regnum));
```
```suggestion
        tcg_gen_discard_tl(target_get_gpr_global(ctx, regnum));
```

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -395,3 +588,1661 @@ static inline void gen_check_cond_branch_target(DisasContext *ctx,
+    cheri_tcg_printf_verbose("cd", "Get reg %d base: %lx\n", regnum, base);
+}
+
+#if CHERI_CAP_BITS == 128 // TODO: make it work for CHERI-64
```
Some CHERI-64 code exists, but it looks like some functions are missing?

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -395,3 +588,1661 @@ static inline void gen_check_cond_branch_target(DisasContext *ctx,
+#ifdef DO_TCG_BOUNDS_CHECKS
+    // We use addr as a tmp, and tmp as address because we can ensure tmp is
+    // local
+    TCGv_i64 local_addr = tcg_temp_local_new_i64();
```
These should all be tl given they're not under AARCH64 defines (only happen to be due to that being the only TCG platform); (and yes, gen_cap_in_bounds expects result to be a tl).

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -155,14 +174,26 @@ static inline bool cap_is_sealed_with_type(const cap_register_t *c)
+    /*
+     * Morello has positive signed otypes.
+     */
+    return cap_get_otype_unsigned(c) > CAP_LAST_NONRESERVED_OTYPE;
```
This doesn't look right, the last non-reserved otype is 2^N-1 on Morello, unless this is a misnomer that should be fixed?

LawrenceEsswood replied:
I think its a misnomer because of the reversal of which end of the type space has special types between morello and other platforms. The name of this field existed prior to morello.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -202,30 +233,41 @@ static inline bool cap_is_unsealed(const cap_register_t *c)
 static inline void cap_set_sealed(cap_register_t *c, uint32_t type)
 {
+#ifdef TARGET_AARCH64
+    type &= CAP_CC(FIELD_OTYPE_MASK_NOT_SHIFTED);
```
When is this needed?

LawrenceEsswood replied:
This might be another case where mips/riscv faulted if type was too large, but morello just masked it. I think I stopped using the helper, but we should still maintain correctness for the helper as it is reasonable to call for morello, and should not break when it does.

jrtc27 replied:
Ok, I read the spec, and Morello will truncate+tag clear in that case. That needs handling here if you want the helper to be correct, otherwise don't do anything here, having a half-baked implementation is the worst option.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -202,30 +233,41 @@ static inline bool cap_is_unsealed(const cap_register_t *c)
 {
     assert(c->cr_tag);
+#ifndef TARGET_AARCH64
     assert(cap_is_sealed_with_type(c));
```
This looks like it should still hold

LawrenceEsswood replied:
Well, now that Morello is less reliant on helpers (doing most things in tcg), there isn't _currently_ a use of this that will hit the assert. However, morello allows unsealing of capabilities with special types using normal unseal operations. I have hit this in the past when using this helper, the asserts just don't make sense for morello.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -202,30 +233,41 @@ static inline bool cap_is_unsealed(const cap_register_t *c)
     assert(c->cr_tag);
+#ifndef TARGET_AARCH64
     assert(cap_is_sealed_with_type(c));
     assert(cap_get_otype_unsigned(c) <= CAP_LAST_NONRESERVED_OTYPE &&
```
This should still hold if you use cap_is_sealed_with_type, or inline the subset needed here

arichardson replied:
Ah yeah, I think I updated the call that asserts here to use a direct otype update.

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -253,16 +344,78 @@ void CHERI_HELPER_IMPL(ccleartag(CPUArchState *env, uint32_t cd, uint32_t cb))
+                         target_ulong addr, uint32_t link_reg,
+                         target_ulong link_pc, uint32_t cjalr_flags)
+{
+    // FIXME: I am concerned that morello will want to jump to sealed things and
```
Are these known now? (And I believe it does, right?)

LawrenceEsswood replied:
It does, and I think it now gets it right. Comment should be removed.

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -1109,15 +1218,30 @@ cheri_tag_prot_clear_or_trap(CPUArchState *env, target_ulong va,
+        (CAP_cc(cap_pesbt_extract_otype)(*pesbt) == CAP_OTYPE_UNSEALED)) {
+        qemu_maybe_log_instr_extra(env,
+                                   "Squashing mutable load related perms\n");
+        *pesbt &= ~cc128_cap_pesbt_encode_perms(
```
```suggestion
        *pesbt &= ~CAP_cc(cap_pesbt_encode_perms)(
```

LawrenceEsswood replied:
I will get rid.

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -1109,15 +1218,30 @@ cheri_tag_prot_clear_or_trap(CPUArchState *env, target_ulong va,
+        qemu_maybe_log_instr_extra(env,
+                                   "Squashing mutable load related perms\n");
+        *pesbt &= ~cc128_cap_pesbt_encode_perms(
+            CC128_PERM_MUTABLE_LOAD | CC128_PERM_STORE_LOCAL |
```
Though there are loads of these scattered throughout TARGET_AARCH64 code that will need changing to CAP_

LawrenceEsswood replied:
Also will get rid.

arichardson committed:
Drop duplicate call to cheri_tag_init() for morello board

The tags are initialized automatically for the default RAM device.

arichardson commented:
rebased.

arichardson committed:
Revert "Remove assert as behaviour is legal"

We should always have either an in-bounds PCC or be in the processes of
raising and invalid PCC exception. In the latter case we don't need to
restore state since tcg should have saved pc prior to calling the raise
pcc exception helper.

This reverts commit 4fee44544754dc15e1ecbd769085a9b7e5a840cb.

arichardson committed:
Rename TARGET_IS_MORELLO to TARGET_MORELLO

As suggested by @jrtc27

LawrenceEsswood committed:
Fix some comments

LawrenceEsswood committed:
Branch and link bug has been fixed in ASL

LawrenceEsswood committed:
Use typedef not #define

LawrenceEsswood committed:
Add assert in store tags helper

LawrenceEsswood committed:
Dont assume CC128

LawrenceEsswood committed:
Turn logical op into else branch

LawrenceEsswood committed:
Slightly better comment / name for capability swap

LawrenceEsswood committed:
Distinguish between signed and unsigned in tcg_cond_string

LawrenceEsswood committed:
Sensible ordering in list of masks

LawrenceEsswood committed:
Fix PAGE_C_BITS define to actually include all bits

LawrenceEsswood committed:
Use canonical bit inclusion pattern

LawrenceEsswood committed:
ifdef out dead code

LawrenceEsswood committed:
Neaten up defines

LawrenceEsswood committed:
Less crazy line brbreak

LawrenceEsswood committed:
Fix cpu get asid

LawrenceEsswood committed:
Revert type change in mte_check_zva

LawrenceEsswood committed:
Make the mpidr bit a property

LawrenceEsswood committed:
Small whitespace fix

LawrenceEsswood committed:
Inline return one for is_a64 for cheri arm

arichardson committed:
Update clang-format config to be closer to upstream code style

LawrenceEsswood committed:
Pass PC to cpu_get_asid

LawrenceEsswood committed:
Make some more TCG work for 32-bit

LawrenceEsswood committed:
Remove stale comments

LawrenceEsswood committed:
Simplify some flow control

LawrenceEsswood committed:
Fix type of placeholder for RISCV target_get_gpr_global

arichardson commented:
I'm happy to merge this now (with a merge commit) and deal with further cleanups after merging.

jrtc27 reviewed (commented):
Other than the comments in this review, I believe the outstanding issues are:

1. Moving INSN_ARM_C64 (or telling me you don't think it should be moved :P)
2. ~~Removing map.h and deferring its inclusion (or discussion of whether it's even the right thing to use) until the PR that needs it~~ I realise I commented on it too...
3. Comments in cap_set_(un)sealed
4. Various clang-format diffs that may lead to conflicts

1-3 should be simple to resolve. 4 is a pain but I don't mind volunteering time to go through the tedium of reformatting to diff-reduce as I'd rather not have those kinds of mass-reformattings mess with git-blame (the whitespace ones matter less due to -w, but -w doesn't filter out things that include changing where line breaks are...).

LawrenceEsswood committed:
Change asserts for cap_set_(un)sealed for AARCH64

arichardson commented:
> Other than the comments in this review, I believe the outstanding issues are:
> 
> 1. Moving INSN_ARM_C64 (or telling me you don't think it should be moved :P)

Sure I can remove that, but then I guess we should also move INSN_ARM_BE32 (and possibly conflict with upstream?).

> 2. ~Removing map.h and deferring its inclusion (or discussion of whether it's even the right thing to use) until the PR that needs it~ I realise I commented on it too...

I think we could just drop that commit for now.

> 3. Comments in cap_set_(un)sealed

Morello let's you unseal reserved types, but that should probably be handled in the helper instead of the cap_set_(un)sealed wrappers. I don't feel strongly either way.

> 4. Various clang-format diffs that may lead to conflicts
> 
> 1-3 should be simple to resolve. 4 is a pain but I don't mind volunteering time to go through the tedium of reformatting to diff-reduce as I'd rather not have those kinds of mass-reformattings mess with git-blame (the whitespace ones matter less due to -w, but -w doesn't filter out things that include changing where line breaks are...).

That would be great, I tried to clean up some of them but it is quite tedious to clean up ones that also include some functional changes.

jrtc27 commented:
> > Other than the comments in this review, I believe the outstanding issues are:
> > 
> > 1. Moving INSN_ARM_C64 (or telling me you don't think it should be moved :P)
> 
> Sure I can remove that, but then I guess we should also move INSN_ARM_BE32 (and possibly conflict with upstream?).

Those can stay, and given they adhere to binutils's encoding convention we should be safe assuming they won't clash with things allocated from 1 up.

LawrenceEsswood commented:
I don't feel strongly on 3 either. Morello does not use cap_set_sealed so its now just assert(0). cap_set_unsealed can have the first assert, as its only caller (the morello equivalent of CCall) is not making the call if its a special type. I am kind of sad that this is the case, I would have liked to see if there was anything fun that could be done with sentries that could use used 'as-is' or as a ccall argument.

LawrenceEsswood commented:
On (2), I don't mind arguing for map if/when it is needed.

jrtc27 committed:
morello: Diff-reduce tables (and match style for new ones)

jrtc27 committed:
tcg: Diff-reduce GEN_ATOMIC_HELPER

jrtc27 committed:
morello: Fix up clang-format directives

One was missing and one was poor style.

arichardson committed:
Rename INSN_ARM_C64 to ARM_DIS_FLAG_C64

This matches RISC-V and avoids using the binutils INSN_* namespace.

arichardson committed:
Reformat translate-a64.c with `IndentCaseBlocks: true`

This reduces the diff to upstream.

arichardson committed:
Simplify cap_is_unsealed()

It is impossible to have an otype outside of the valid range with the
current compression schemes. Now that we extract the otype from the
pesbt bits it's also impossible for cap_get_otype_unsigned() to return
such a value.

arichardson committed:
Clean up Morello ifdefs in cap_set_sealed/cap_set_unsealed

These helpers are not called by the TCG generated for Morello.

arichardson committed:
Add a cap_unseal_reserved_otype() helper

This removes some CAP_cc usages from the Morello helpers.

arichardson committed:
Cleanup cap_is_sealed_with_type()

And factor out a new cap_otype_is_reserved() helper function.

arichardson committed:
Simplify assertions in cap_set_sealed() and cap_set_unsealed()

After the latest refactoring we can now drop the TARGET_MORELLO branch.

arichardson committed:
Fix value of CC{64,128}_LAST_NONRESERVED_OTYPE

Most uses appear to assume that LAST_NONRESERVED_OTYPE is a valid
non-reserved otype rather than the first reserved one. However, all of
those checks are incorrect for Morello, so we should really remove
this macro.

arichardson committed:
Stop using CAP_LAST_NONRESERVED_OTYPE

Performing comparisons against this constant is likely to cause bugs
since Morello allocates otypes from the bottom but MIPS/RISC-V allocate
from the top, so any less than/greater than checks are wrong for at
least one architecture.

arichardson committed:
[cheri-compressed-cap] Replace FIRST/LAST otype with MIN/MAX

This makes it clear that we are referring the the MIN/MAX numerical value.
Previously it was -1 to -16 for MIPS/RISC-V which meant that my
implementation of cap_otype_is_reserved() was broken since the check
`otype > 0x3ffff && otype <= 0x3fff0` is never true.

arichardson committed:
[cheri-compressed-cap] Assert cap_pesbt_deposit_X arguments are in range

This does not corrupt other fields, but they will be truncated silently.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -133,10 +133,19 @@ static inline cap_length_t cap_get_top_full(const cap_register_t *c)
 static inline bool cap_otype_is_reserved(target_ulong otype)
 {
+    cheri_debug_assert((target_long)otype >= 0 &&
+                       "Should not only be used with sign-extended otypes!");
```
"Should not be ..."

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -133,10 +133,19 @@ static inline cap_length_t cap_get_top_full(const cap_register_t *c)
+    cheri_debug_assert((target_long)otype >= 0 &&
+                       "Should not only be used with sign-extended otypes!");
+    cheri_debug_assert(otype <= CAP_MAX_REPRESENTABLE_OTYPE &&
+                       "Should only be called for in-range otypes!");
```
Doesn't this encompass the above assertion?

arichardson replied:
Yes it does. I had another stupid bug so I added the first assertion. Will drop a re-push.

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -445,7 +452,7 @@ void CHERI_HELPER_IMPL(cinvoke(CPUArchState *env, uint32_t code_regnum,
         raise_cheri_exception(env, CapEx_SealViolation, data_regnum);
     } else if (cap_get_otype_unsigned(code_cap) != cap_get_otype_unsigned(data_cap) ||
-               cap_get_otype_unsigned(code_cap) > CAP_LAST_NONRESERVED_OTYPE) {
+               !cap_otype_is_nonreserved(cap_get_otype_unsigned(code_cap))) {
```
!cap_is_sealed_with_type?

arichardson replied:
yes that is better 🤦

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -314,14 +314,21 @@ target_ulong CHERI_HELPER_IMPL(cgettype(CPUArchState *env, uint32_t cb))
-        cheri_debug_assert(otype <= CAP_LAST_NONRESERVED_OTYPE);
+        cheri_debug_assert(
+            cap_is_sealed_with_type(cbp) &&
+            "positive return values are used for non-reserved otypes");
```
non-negative

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -133,10 +133,19 @@ static inline cap_length_t cap_get_top_full(const cap_register_t *c)
 
+static inline bool cap_otype_is_nonreserved(target_ulong otype)
+{
+    return otype < CAP_MAX_REPRESENTABLE_OTYPE && !cap_otype_is_reserved(otype);
```
<=? Otherwise max otype isn't handled correctly on Morello (but happens to be on CHERI due to that being a reserved otype)

jrtc27 replied:
Also, cap_otype_is_reserved asserting but cap_otype_is_reserved not asserting is a bit confusing, it's really tempting to turn the !cap_otype_is_nonreserved calls into cap_otype_is_reserved, but can't due to the difference between the two

jrtc27 reviewed (approved):
Let's hope this one didn't break anything... 🤞
